File: objc/block/BlockRetainCycleObjc/BlockRetainCycleObjc/ReporterView.m
//
//  ReporterView.m
//  BlockRetainCycleObjc
//
//  Created by zhen gong on 6/12/17.
//  Copyright © 2017 zhen gong. All rights reserved.
//  https://swiftcafe.io/2017/02/02/weak-block/

#import "ReporterView.h"

@interface ReporterView()

@end

@implementation ReporterView

- (instancetype)init
{
    self = [super init];
    if (self) {
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(handleNotification)
                                                     name:@"Notification"
                                                   object:nil];
    }
    return self;
}

-(void)handleNotification
{
    DLog(@"Notification Received");
}

-(void)foo
{
    DLog(@"Foo is called.")
}

-(void)block:(void (^)(void))completionBlock {
    // doBlock is strong type
    DLog(@"ReporterView call block method");
    self.doBlock = completionBlock;
    self.doBlock();
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    DLog(@"dealloc");
}

@end

File: objc/block/BlockRetainCycleSwift/BlockRetainCycleSwift/ReporterView.swift
//
//  ReporterView.swift
//  BlockRetainCycleSwift
//
//  Created by zhen gong on 6/12/17.
//  Copyright © 2017 zhen gong. All rights reserved.
//

import UIKit

class ReporterView: UIView {
    
    var doBlock:(() -> ())?
    
    var bobClosure:(() -> ())?
    var bobName: String = {
        return "Bob"
    }()
    
    init() {
        super.init(frame: CGRect.zero)
        self.bobClosure = {
//            print("bobClosure \(self.bobName)")
        }
        NotificationCenter.default.addObserver(self, selector: #selector(handleNotification), name: NSNotification.Name(rawValue: "Notification"), object: nil)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError()
    }
    
    
    func block(closureBlock:@escaping ((Void) -> (Void))) {
        self.doBlock = closureBlock
        if let doBlock = self.doBlock {
            print("ReporterView call block method:\(self.bobName)")
            doBlock()
        }
        
    }
    
    func foo() {
        NSLog("Foo is called")
    }
    
    func handleNotification() {
        NSLog("Notification Received")
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
        NSLog("dealloc")
    }

}
